hadoop-hdds/docs/content/design/container-reconciliation.md

```diff
@@ -0,0 +1,246 @@
+- `getContainerHashes(containerID)`
+    - A datanode API that returns the merkle tree for a given container. The proto structure would be similar to that outlined in [Storage](#storage)
+
+## Events
+
+This section defines how container reconciliation will function as events occur in the system.
+
+### Reconiliation Events
+
+ These events occur as part of container reconciliation in the happy path case.
+
+#### On Data Write
+
+- No change to existing datanode operations. Existing chunk checksums passed and validated from the client will be persisted to RocksDB.
+
+#### On Container Close
+
+- Container checksum is calculated synchonously from existing checksums. This calculation must finish before the close is acked back to SCM.
+    - **Invariant**: All closed containers have a checksum even in the the case of restarts and failures because SCM will retry the close command if it does not receive an ack.
+      - Null handling should still be in place for containers created before this feature, and to guard against bugs.
+
+#### On Container Scan
+
+1. Scanner reads chunks and checks checksums in RocksDB.
+2. Scanner updates Merkle tree to match the current contents of the disk. It does not update RocksDB.
+    - Update Merkle tree checksums in a rolling fashion per chunk and block as scanning progresses.
+    -  Do not let the scanner overwrite the RocksDB value, otherwise we cannot locally detect corruption anymore.
+        - This is not a deal breaker since SCM would see it on the next full container report, but it may slow down bitrot detection.
+3. If scanner detects local corruption (RocksDB hash does not match hash calculated from disk), marks container unhealthy and sends an incrmeental container report to SCM with its calculated container hash.
+    - The Merkle Tree is still updated to reflect the current contents of the disk in this case.
+
+#### On Container import
+
+- Schedule on-demand scan after the container import completes.
+    - There may be cases where we need to import a slightly corrupt container since it is our best option. Therefore, do not do a synchronous scan of container and fail the import if corruption is detected.
+    - Instead, schedule the imported container for on-demand scanning to be sure SCM knows of corruption if it does not already.
+    - **Optimization**: Clear last scanned timestamp on container import to make the background scanner get to it faster if datanode restarts before the on-demand scan completes.
+
+#### On `getContainerHashes`
+
+- Return the current Merkle Tree of the specified container.
+    - The tree will be calculated from existing chunk checksums if it is not present. However, this should not happen in the majority of cases since it will be calculated on container close.
+- If container not closed, return error.
+- If container not found, return error.
+
+- The merkle tree returned by this call will always be consistent.
```

Comment:
The proto file could be corrupted.  We need to checksum it. :)

Reply:
Good point. we can also store a hash within the tree that represents the has of the content excluding the hash itself.

Reply:
Will cover this in the implementation for the Merkle Tree